File: example/transparent_cartesian_teleop_lan.cpp
/**
 * @example transparent_cartesian_teleop_lan.cpp
 * @brief Example usage of Transparent Cartesian teleoperation under Local Area Network for
 * controlling a follower robot using a leader robot with transparent force feedback. Supports both
 * keyboard and digital input engage/disengage signal reading, with various axes lock modes, force
 * scaling, max contact wrench setting, and teleop status query, etc.
 * @note This program is provided only as an example. Users must adapt it to their own application
 * requirements, safety procedures, and software architecture before deployment.
 * @copyright Copyright (C) 2016-2026 Flexiv Ltd. All Rights Reserved.
 * @author Flexiv
 */

#include <flexiv/tdk/data.hpp>
#include <flexiv/tdk/transparent_cartesian_teleop_lan.hpp>

#include <getopt.h>
#include <array>
#include <atomic>
#include <chrono>
#include <cmath>
#include <functional>
#include <iostream>
#include <optional>
#include <string>
#include <thread>
#include <vector>

namespace {
/** Nullspace to a preferred posture */
const std::vector<double> kPreferredJntPos
    = {60 * M_PI / 180.0, -60 * M_PI / 180.0, -85 * M_PI / 180.0, 115 * M_PI / 180.0,
        70 * M_PI / 180.0, 0 * M_PI / 180.0, 0 * M_PI / 180.0};

/** Nullspace to Home posture */
const std::vector<double> kHomeJntPos = {0 * M_PI / 180.0, -40 * M_PI / 180.0, 0 * M_PI / 180.0,
    90 * M_PI / 180.0, 0 * M_PI / 180.0, 40 * M_PI / 180.0, 0 * M_PI / 180.0};

/** Maximum contact wrench for soft contact */
const std::array<double, flexiv::tdk::kCartDoF> kDefaultMaxContactWrench
    = {5.0, 5.0, 5.0, 40.0, 40.0, 40.0};

/** Robot pair index used by this example */
constexpr unsigned int kIdx = 0;

/** Single-arm joint group controlled by this example */
constexpr auto kJointGroup = flexiv::rdk::JointGroup::ARM_1;

/** Atomic signal to stop console and DI reading tasks */
std::atomic<bool> g_running {true};

void LogInfo(const std::string& msg)
{
    std::cout << "[info] " << msg << std::endl;
}

void LogWarn(const std::string& msg)
{
    std::cerr << "[warn] " << msg << std::endl;
}

void LogError(const std::string& msg)
{
    std::cerr << "[error] " << msg << std::endl;
}

void PrintTeleopStatus(const flexiv::tdk::TeleopStatus& status)
{
    LogInfo("Teleop status: initialized=" + std::to_string(status.initialized)
            + " started=" + std::to_string(status.started)
            + " engaged=" + std::to_string(status.engaged)
            + " stopped=" + std::to_string(status.stopped)
            + " fault=" + std::to_string(status.fault)
            + " motion_restricted=" + std::to_string(status.motion_restricted));
    if (status.primary.code == flexiv::tdk::TeleopIssueCode::NONE) {
        LogInfo("No teleop restriction. Safe to continue.");
        return;
    }
    const auto& issue = status.primary;
    LogWarn(std::string("[")
            + flexiv::tdk::TeleopIssueLevelStr[static_cast<size_t>(issue.level)] + "] "
            + issue.title + " | " + issue.description + " | " + issue.suggestion);
    for (const auto& extra : status.issues) {
        if (extra.code == issue.code && extra.side == issue.side
            && extra.joint_index == issue.joint_index) {
            continue;
        }
        LogWarn(std::string("  also: [")
                + flexiv::tdk::TeleopIssueCodeStr[static_cast<size_t>(extra.code)] + "] "
                + extra.title);
    }
}

} // namespace

void PrintHelp()
{
    // clang-format off
    std::cout<<"Invalid program arguments!"<<std::endl;
    std::cout<<"     -l     [necessary] serial number of leader robot."<<std::endl;
    std::cout<<"     -f     [necessary] serial number of follower robot."<<std::endl;
    std::cout<<"     -D     [optional] Enable Digital Input reading task." << std::endl;
    std::cout<<"Usage: sudo ./transparent_cartesian_teleop_lan [-l leader_robot_sn] [-f follower_robot_sn] [-D]"<<std::endl;
    // clang-format on
}

const struct option kLongOptions[] = {
    // clang-format off
    {"leader SN",                                       required_argument,  0, 'l'},
    {"follower SN",                                     required_argument,  0, 'f'},
    {"enable digital input",                            no_argument,        0, 'D'},
    {0,                                                                 0,  0,  0 }
    // clang-format on
};

/**
 * @brief Task for monitoring DI signals and engaging teleop.
 */
void ReadDigitalInputTask(flexiv::tdk::TransparentCartesianTeleopLAN& teleop)
{
    while (g_running.load() && !teleop.any_fault()) {
        try {
            teleop.Engage(kIdx, kJointGroup, teleop.digital_inputs(kIdx).first[0]);
        } catch (const std::exception& e) {
            LogError(std::string("Exception in ReadDigitalInputTask: ") + e.what());
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }
    LogInfo("ReadDigitalInputTask exiting.");
    return;
}
/**
 * @brief Task for calling TransparentCartesianTeleopLAN functions from console.
 */
void ConsoleTask(flexiv::tdk::TransparentCartesianTeleopLAN& teleop)
{
    auto PrintCommandMenu = []() {
        std::cout << R"(
  --- Axis Lock ---
    x/y/z    : Toggle translation lock in WORLD coord (X/Y/Z)
    q/w/e    : Toggle orientation lock in WORLD coord (Rx/Ry/Rz)
    X/Y/Z    : Toggle translation lock in TCP coord (X/Y/Z)
    Q/W/E    : Toggle orientation lock in TCP coord (Rx/Ry/Rz)

  --- Teleop Engagement ---
    r        : Engage teleop
    R        : Disengage teleop

  --- Wrench Feedback Scaling ---
    t        : Set wrench feedback scaling to 0.5
    T        : Set wrench feedback scaling to 2.0
    c        : Reset wrench feedback scaling to 1.0 default value

  --- Axis Lock Presets ---
    u        : Unlock all axes (TCP coord)
    U        : Lock all axes (TCP coord)

  --- Null Space Posture ---
    i/I      : Set local nullspace to Preferred/Home posture
    o/O      : Set remote nullspace to Preferred/Home posture

  --- Max Contact Wrench ---
    p        : Set default remote max contact wrench

  --- Repulsive Force ---
    a        : Set repulsive force to {5, 0, 0}
    A        : Set repulsive force to {-5, 0, 0}
    C        : Clear force 

  --- Start/Stop ---
    b        : Stop teleop
    B        : Start teleop


  --- Is teleop stopped or not ---
    s        : Query if teleop stopped or not

  --- Teleop status ---
    h        : Print why teleop is restricted / paused and what to do next

  --- Help ---
    Any other key to show this help menu
        )" << std::endl;
    };

    flexiv::tdk::AxisLock cmd;
    teleop.GetAxisLockState(kIdx, kJointGroup, cmd);

    while (g_running.load() && !teleop.any_fault()) {

        std::string userInput {};

        // Get user input
        std::getline(std::cin, userInput);

        if (userInput.empty()) {
            LogWarn("Empty command!");
            PrintCommandMenu();
            continue;
        }

        try {
            switch (userInput[0]) {
                case 'x':
                    cmd.lock_trans_axis[0] = !cmd.lock_trans_axis[0];
                    cmd.coord = flexiv::tdk::CoordType::COORD_WORLD;
                    teleop.SetAxisLockCmd(kIdx, kJointGroup, cmd);
                    break;
                case 'y':
                    cmd.lock_trans_axis[1] = !cmd.lock_trans_axis[1];
                    cmd.coord = flexiv::tdk::CoordType::COORD_WORLD;
                    teleop.SetAxisLockCmd(kIdx, kJointGroup, cmd);
                    break;
                case 'z':
                    cmd.lock_trans_axis[2] = !cmd.lock_trans_axis[2];
                    cmd.coord = flexiv::tdk::CoordType::COORD_WORLD;
                    teleop.SetAxisLockCmd(kIdx, kJointGroup, cmd);
                    break;
                case 'q':
                    cmd.lock_ori_axis[0] = !cmd.lock_ori_axis[0];
                    cmd.coord = flexiv::tdk::CoordType::COORD_WORLD;
                    teleop.SetAxisLockCmd(kIdx, kJointGroup, cmd);
                    break;
                case 'w':
                    cmd.lock_ori_axis[1] = !cmd.lock_ori_axis[1];
                    cmd.coord = flexiv::tdk::CoordType::COORD_WORLD;
                    teleop.SetAxisLockCmd(kIdx, kJointGroup, cmd);
                    break;
                case 'e':
                    cmd.lock_ori_axis[2] = !cmd.lock_ori_axis[2];
                    cmd.coord = flexiv::tdk::CoordType::COORD_WORLD;
                    teleop.SetAxisLockCmd(kIdx, kJointGroup, cmd);
                    break;

                case 'X':
                    cmd.lock_trans_axis[0] = !cmd.lock_trans_axis[0];
                    cmd.coord = flexiv::tdk::CoordType::COORD_TCP;
                    teleop.SetAxisLockCmd(kIdx, kJointGroup, cmd);
                    break;
                case 'Y':
                    cmd.lock_trans_axis[1] = !cmd.lock_trans_axis[1];
                    cmd.coord = flexiv::tdk::CoordType::COORD_TCP;
                    teleop.SetAxisLockCmd(kIdx, kJointGroup, cmd);
                    break;
                case 'Z':
                    cmd.lock_trans_axis[2] = !cmd.lock_trans_axis[2];
                    cmd.coord = flexiv::tdk::CoordType::COORD_TCP;
                    teleop.SetAxisLockCmd(kIdx, kJointGroup, cmd);
                    break;
                case 'Q':
                    cmd.lock_ori_axis[0] = !cmd.lock_ori_axis[0];
                    cmd.coord = flexiv::tdk::CoordType::COORD_TCP;
                    teleop.SetAxisLockCmd(kIdx, kJointGroup, cmd);
                    break;
                case 'W':
                    cmd.lock_ori_axis[1] = !cmd.lock_ori_axis[1];
                    cmd.coord = flexiv::tdk::CoordType::COORD_TCP;
                    teleop.SetAxisLockCmd(kIdx, kJointGroup, cmd);
                    break;
                case 'E':
                    cmd.lock_ori_axis[2] = !cmd.lock_ori_axis[2];
                    cmd.coord = flexiv::tdk::CoordType::COORD_TCP;
                    teleop.SetAxisLockCmd(kIdx, kJointGroup, cmd);
                    break;

                case 'r':
                    teleop.Engage(kIdx, kJointGroup, true);
                    break;
                case 'R':
                    teleop.Engage(kIdx, kJointGroup, false);
                    break;

                case 't':
                    teleop.SetWrenchFeedbackScalingFactor(kIdx, kJointGroup, 0.5);
                    break;
                case 'T':
                    teleop.SetWrenchFeedbackScalingFactor(kIdx, kJointGroup, 2);
                    break;

                case 'u':
                    cmd.lock_ori_axis = {false, false, false};
                    cmd.lock_trans_axis = {false, false, false};
                    cmd.coord = flexiv::tdk::CoordType::COORD_TCP;
                    teleop.SetAxisLockCmd(kIdx, kJointGroup, cmd);
                    break;
                case 'U':
                    cmd.lock_ori_axis = {true, true, true};
                    cmd.lock_trans_axis = {true, true, true};
                    cmd.coord = flexiv::tdk::CoordType::COORD_TCP;
                    teleop.SetAxisLockCmd(kIdx, kJointGroup, cmd);
                    break;
                case 'i':
                    teleop.SetLeaderNullSpacePosture(kIdx, kJointGroup, kPreferredJntPos);
                    break;
                case 'I':
                    teleop.SetLeaderNullSpacePosture(kIdx, kJointGroup, kHomeJntPos);
                    break;
                case 'o':
                    teleop.SetFollowerNullSpacePosture(kIdx, kJointGroup, kPreferredJntPos);
                    break;
                case 'O':
                    teleop.SetFollowerNullSpacePosture(kIdx, kJointGroup, kHomeJntPos);
                    break;
                case 'p':
                    teleop.SetFollowerMaxContactWrench(kIdx, kJointGroup, kDefaultMaxContactWrench);
                    break;

                case 'a':
                    teleop.SetRepulsiveForce(kIdx, kJointGroup, {5, 0, 0});
                    break;
                case 'A':
                    teleop.SetRepulsiveForce(kIdx, kJointGroup, {-5, 0, 0});
                    break;
                case 'b':
                    teleop.Stop();
                    break;
                case 'B':
                    teleop.Init();
                    teleop.Start();
                    break;
                case 'c':
                    teleop.SetWrenchFeedbackScalingFactor(kIdx, kJointGroup, 1);
                    break;
                case 'C':
                    teleop.SetRepulsiveForce(kIdx, kJointGroup, {0, 0, 0});
                    break;
                case 's':
                    LogInfo(teleop.stopped(kIdx) ? "Teleop pair 0 stopped" : "Teleop pair 0 started");
                    break;
                case 'h':
                    PrintTeleopStatus(teleop.GetTeleopStatus(kIdx, kJointGroup));
                    break;
                default:
                    LogWarn("Invalid command!");
                    PrintCommandMenu();
                    break;
            }
        } catch (const std::exception& e) {
            LogError(std::string("Exception in ConsoleTask: ") + e.what());
            g_running.store(false);
            return;
        }
    }
    LogInfo("Console thread exiting.");
    return;
}

int main(int argc, char* argv[])
{
    std::string follower_sn {};
    std::string leader_sn {};
    int opt = 0;
    int longIndex = 0;
    bool enable_digital_input = false;

    while ((opt = getopt_long_only(argc, argv, "f:l:D", kLongOptions, &longIndex)) != -1) {
        switch (opt) {
            case 'f':
                follower_sn = std::string(optarg);
                break;
            case 'l':
                leader_sn = std::string(optarg);
                break;
            case 'D':
                enable_digital_input = true;
                break;
            default:
                PrintHelp();
                return 1;
        }
    }
    if (leader_sn.empty() || follower_sn.empty()) {
        PrintHelp();
        return 1;
    }
    try {

        // Create teleop control interface
        flexiv::tdk::TransparentCartesianTeleopLAN tctl({{leader_sn, follower_sn}});

        // Run initialization sequence
        tctl.Init();

        // Start control loop
        tctl.Start();

        // Start console and pedal input threads
        std::thread console_thread(std::bind(ConsoleTask, std::ref(tctl)));

        // Start pedal_thread based on the new flag or the role
        std::optional<std::thread> pedal_thread;

        if (enable_digital_input) {
            LogInfo("Starting ReadDigitalInputTask thread.");
            pedal_thread.emplace(ReadDigitalInputTask, std::ref(tctl));
        } else {
            LogInfo("ReadDigitalInputTask thread NOT started (-D flag not provided).");
        }

        // Wait for threads to finish
        console_thread.join();

        // Stop all threads, notify other threads exit
        g_running.store(false);

        // Wait for digital reading task exit
        if (pedal_thread && pedal_thread->joinable()) {
            pedal_thread->join();
        }

        // Exit high transparency teleop
        tctl.Stop();

    } catch (const std::exception& e) {
        LogError(e.what());
        return 1;
    }

    return 0;
}

File: example_py/transparent_cartesian_teleop_lan.py
#!/usr/bin/env python

"""
transparent_cartesian_teleop_lan.py

Example usage of Transparent Cartesian teleoperation under Local Area Network,
controlling a follower robot using a leader robot with transparent force feedback.
Supports both keyboard and digital input engage/disengage signal reading, 
with various axes lock modes, force scaling, and max contact wrench setting, etc.

"""

__copyright__ = "Copyright (C) 2016-2025 Flexiv Ltd. All Rights Reserved."
__author__ = "Flexiv"

import argparse
import threading
import time
import sys
import spdlog
import math
from typing import List, Optional, Dict, Callable

# pip install flexivtdk
import flexivtdk  


# Helper to convert degree lists to radians
def _deg2rad_list(deg_list):
    return [math.radians(d) for d in deg_list]

# Example of null-space postures
kPreferredJntPos = _deg2rad_list([60, -60, -85, 115, 70, 0, 0])
kHomeJntPos = _deg2rad_list([0, -40, 0, 90, 0, 40, 0])

# Example of max contact wrench setting
kDefaultMaxContactWrench = [25.0, 25.0, 25.0, 40.0, 40.0, 40.0]

# Global thread-safe stop event
_stop_event = threading.Event()

# Logger setup
logger = spdlog.ConsoleLogger("Example")

class TeleoperationController:
    """Encapsulates teleoperation functionality for better organization and maintainability."""
    
    def __init__(self, teleop: flexivtdk.TransparentCartesianTeleopLAN):
        self.teleop = teleop
        self.index = 0
        self.cmd = self._get_initial_axis_lock_cmd()
        self._command_map = self._create_command_map()
        self._menu = self._create_menu()
    
    def _get_initial_axis_lock_cmd(self):
        """Initialize axis lock command object."""
        try:
            return self.teleop.GetAxisLockState(self.index)
        except Exception as e:
            # Fallback if GetAxisLockState overload returns differently
            cmd = flexivtdk.AxisLock()
            cmd.lock_trans_axis = [False, False, False]
            cmd.lock_ori_axis = [False, False, False]
            cmd.coord = flexivtdk.CoordType.TCP
            logger.warning(f"Failed to get initial axis lock state, using default: {e}")
            return cmd
    
    def _create_command_map(self) -> Dict[str, Callable]:
        """Create a mapping of keyboard commands to their corresponding methods."""
        return {
            # Translation locks WORLD coord
            'x': lambda: self._toggle_axis_lock(0, 'trans', flexivtdk.CoordType.WORLD),
            'y': lambda: self._toggle_axis_lock(1, 'trans', flexivtdk.CoordType.WORLD),
            'z': lambda: self._toggle_axis_lock(2, 'trans', flexivtdk.CoordType.WORLD),
            # Orientation locks WORLD coord
            'q': lambda: self._toggle_axis_lock(0, 'ori', flexivtdk.CoordType.WORLD),
            'w': lambda: self._toggle_axis_lock(1, 'ori', flexivtdk.CoordType.WORLD),
            'e': lambda: self._toggle_axis_lock(2, 'ori', flexivtdk.CoordType.WORLD),
            # Translation locks TCP coord
            'X': lambda: self._toggle_axis_lock(0, 'trans', flexivtdk.CoordType.TCP),
            'Y': lambda: self._toggle_axis_lock(1, 'trans', flexivtdk.CoordType.TCP),
            'Z': lambda: self._toggle_axis_lock(2, 'trans', flexivtdk.CoordType.TCP),
            # Orientation locks TCP coord
            'Q': lambda: self._toggle_axis_lock(0, 'ori', flexivtdk.CoordType.TCP),
            'W': lambda: self._toggle_axis_lock(1, 'ori', flexivtdk.CoordType.TCP),
            'E': lambda: self._toggle_axis_lock(2, 'ori', flexivtdk.CoordType.TCP),
            # Teleop engage/disengage
            'r': lambda: self._safe_engage(True),
            'R': lambda: self._safe_engage(False),
            # Wrench feedback scaling
            't': lambda: self._safe_set_wrench_feedback_scaling(0.5),
            'T': lambda: self._safe_set_wrench_feedback_scaling(2.0),
            'c': lambda: self._safe_set_wrench_feedback_scaling(1.0),
            # Presets: unlock/lock all axes in TCP coord
            'u': self._unlock_all_axes,
            'U': self._lock_all_axes,
            # Null-space postures
            'i': lambda: self._safe_set_leader_nullspace(kPreferredJntPos),
            'I': lambda: self._safe_set_leader_nullspace(kHomeJntPos),
            'o': lambda: self._safe_set_follower_nullspace(kPreferredJntPos),
            'O': lambda: self._safe_set_follower_nullspace(kHomeJntPos),
            # Max contact wrench (follower)
            'p': lambda: self._safe_set_follower_max_contact_wrench(kDefaultMaxContactWrench),
            # Start/stop teleop
            'b': self._stop_teleop,
            'B': self._start_teleop,
            # Check teleop stopped state
            's': self._check_teleop_state,
        }
    
    def _create_menu(self) -> str:
        """Create the command menu string."""
        return """
  --- Axis Lock ---
    x/y/z    : Toggle translation lock in WORLD coord (X/Y/Z)
    q/w/e    : Toggle orientation lock in WORLD coord (Rx/Ry/Rz)
    X/Y/Z    : Toggle translation lock in TCP coord (X/Y/Z)
    Q/W/E    : Toggle orientation lock in TCP coord (Rx/Ry/Rz)

  --- Teleop Engagement ---
    r        : Engage teleop
    R        : Disengage teleop

  --- Wrench Feedback Scaling ---
    t        : Set wrench feedback scaling to 0.5
    T        : Set wrench feedback scaling to 2.0
    c        : Reset wrench feedback scaling to 1.0 default value

  --- Axis Lock Presets ---
    u        : Unlock all axes (TCP coord)
    U        : Lock all axes (TCP coord)

  --- Null Space Posture ---
    i/I      : Set local nullspace to Preferred/Home posture
    o/O      : Set remote nullspace to Preferred/Home posture

  --- Max Contact Wrench ---
    p        : Set default remote max contact wrench

  --- Start/Stop ---
    b        : Stop teleop
    B        : Start teleop

  --- Is teleop stopped or not ---
    s        : Query if teleop stopped or not

  --- Help ---
    Any other key to show this help menu
    """
    
    def _toggle_axis_lock(self, axis_index: int, lock_type: str, coord_type: flexivtdk.CoordType):
        """Toggle axis lock for the specified axis and type."""
        try:
            if lock_type == 'trans':
                self.cmd.lock_trans_axis[axis_index] = not self.cmd.lock_trans_axis[axis_index]
            elif lock_type == 'ori':
                self.cmd.lock_ori_axis[axis_index] = not self.cmd.lock_ori_axis[axis_index]
            
            self.cmd.coord = coord_type
            self.teleop.SetAxisLockCmd(self.index, self.cmd)
            logger.info(f"Axis lock toggled: {lock_type}[{axis_index}] = {not (self.cmd.lock_trans_axis[axis_index] if lock_type == 'trans' else self.cmd.lock_ori_axis[axis_index])}, coord = {coord_type}")
        except Exception as e:
            logger.error(f"Failed to toggle axis lock: {e}")
    
    def _unlock_all_axes(self):
        """Unlock all axes in TCP coordinate system."""
        try:
            self.cmd.lock_ori_axis = [False, False, False]
            self.cmd.lock_trans_axis = [False, False, False]
            self.cmd.coord = flexivtdk.CoordType.TCP
            self.teleop.SetAxisLockCmd(self.index, self.cmd)
            logger.info("All axes unlocked")
        except Exception as e:
            logger.error(f"Failed to unlock all axes: {e}")
    
    def _lock_all_axes(self):
        """Lock all axes in TCP coordinate system."""
        try:
            self.cmd.lock_ori_axis = [True, True, True]
            self.cmd.lock_trans_axis = [True, True, True]
            self.cmd.coord = flexivtdk.CoordType.TCP
            self.teleop.SetAxisLockCmd(self.index, self.cmd)
            logger.info("All axes locked")
        except Exception as e:
            logger.error(f"Failed to lock all axes: {e}")
    
    def _start_teleop(self):
        """Initialize and start teleoperation."""
        try:
            # Check if already initialized to avoid re-initialization
            if not hasattr(self, '_initialized') or not self._initialized:
                self.teleop.Init()
                self._initialized = True
            self.teleop.Start()
            logger.info("Teleop started")
        except Exception as e:
            logger.error(f"Failed to start teleop: {e}")
    
    def _stop_teleop(self):
        """Stop teleoperation and set the stop event."""
        try:
            self.teleop.Stop()
            _stop_event.set()
            logger.info("Teleop stopped")
        except Exception as e:
            logger.error(f"Failed to stop teleop: {e}")
            _stop_event.set()
    
    def _check_teleop_state(self):
        """Check and log the teleoperation state."""
        try:
            stopped = self.teleop.stopped(self.index)
            logger.info("Teleop pair {} {}", self.index, "stopped" if stopped else "started")
        except Exception as e:
            logger.error(f"Failed to check teleop state: {e}")
    
    def _safe_engage(self, engage: bool):
        """Safely engage or disengage teleop with error handling."""
        try:
            self.teleop.Engage(self.index, engage)
            logger.info(f"Teleop {'engaged' if engage else 'disengaged'}")
        except Exception as e:
            logger.error(f"Failed to {'engage' if engage else 'disengage'} teleop: {e}")
    
    def _safe_set_wrench_feedback_scaling(self, factor: float):
        """Safely set wrench feedback scaling with error handling."""
        try:
            self.teleop.SetWrenchFeedbackScalingFactor(self.index, factor)
            logger.info(f"Wrench feedback scaling factor set to {factor}")
        except Exception as e:
            logger.error(f"Failed to set wrench feedback scaling: {e}")
    
    def _safe_set_leader_nullspace(self, posture: List[float]):
        """Safely set leader nullspace posture with error handling."""
        try:
            self.teleop.SetLeaderNullSpacePosture(self.index, posture)
            logger.info("Leader nullspace posture set")
        except Exception as e:
            logger.error(f"Failed to set leader nullspace posture: {e}")
    
    def _safe_set_follower_nullspace(self, posture: List[float]):
        """Safely set follower nullspace posture with error handling."""
        try:
            self.teleop.SetFollowerNullSpacePosture(self.index, posture)
            logger.info("Follower nullspace posture set")
        except Exception as e:
            logger.error(f"Failed to set follower nullspace posture: {e}")
    
    def _safe_set_follower_max_contact_wrench(self, wrench: List[float]):
        """Safely set follower max contact wrench with error handling."""
        try:
            self.teleop.SetFollowerMaxContactWrench(self.index, wrench)
            logger.info("Follower max contact wrench set")
        except Exception as e:
            logger.error(f"Failed to set follower max contact wrench: {e}")
    
    def handle_command(self, user_input: str) -> bool:
        """Handle a single user command."""
        if not user_input:
            print(self._menu)
            return True
        
        ch = user_input[0]
        if ch in self._command_map:
            try:
                self._command_map[ch]()
                return True
            except Exception as e:
                logger.error(f"Exception executing command '{ch}': {e}")
                return False
        else:
            logger.info(self._menu)
            return True


# read digital input and engage/disengage teleop accordingly
def read_digital_input_task(teleop: flexivtdk.TransparentCartesianTeleopLAN):
    idx = 0
    while not _stop_event.is_set():
        try:
            # digital_inputs for LAN returns tuple (leader_inputs, follower_inputs)
            di_pair = teleop.digital_inputs(idx)
            leader_di, follower_di = di_pair
            # use leader's first DI port as engage/disengage signal
            if leader_di and len(leader_di) > 0:
                engage_state = bool(leader_di[0])
                teleop.Engage(idx, engage_state)
                logger.info(f"Digital input engage state: {engage_state}")
        except Exception as e:
            logger.error(f"Exception in ReadDigitalInputTask: {e}")
            _stop_event.set()
            return
        time.sleep(0.01)
    logger.info("ReadDigitalInputTask exiting.")


# console task to read user inputs and send commands accordingly
def console_task(teleop: flexivtdk.TransparentCartesianTeleopLAN):
    controller = TeleoperationController(teleop)
    print(controller._menu)
    
    while not _stop_event.is_set():
        try:
            user_input = input("> ").strip()
        except (EOFError, KeyboardInterrupt):
            logger.info("Console exiting by user interrupt.")
            _stop_event.set()
            break
        
        if not controller.handle_command(user_input):
            _stop_event.set()
            break

    logger.info("Console thread exiting.")
    return


def parse_args(argv: Optional[List[str]] = None) -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Transparent Cartesian Teleop LAN example")
    parser.add_argument("-l", "--leader_sn", required=True, help="serial number of leader robot")
    parser.add_argument("-f", "--follower_sn", required=True, help="serial number of follower robot")
    parser.add_argument("-A", "--lan-ip", action="append", help="lan interface ip whitelist", default=[])
    parser.add_argument("-D", "--enable-digital-input", action="store_true", help="enable digital input reading task")
    return parser.parse_args(argv)


def main(argv: Optional[List[str]] = None):
    # Program Setup
    # ==============================================================================================
    # Parse arguments
    args = parse_args(argv)
    robot_pairs = [(args.leader_sn, args.follower_sn)]
    try:
        # TDK Initialization
        # ==========================================================================================
        # Instantiate teleop interface
        teleop = flexivtdk.TransparentCartesianTeleopLAN(robot_pairs, args.lan_ip)
        teleop.Init()
        teleop.Start()
        logger.info("Teleop started.")

        # Start console task thread
        console_thr = threading.Thread(target=console_task, args=(teleop,), daemon=True)
        console_thr.start()
        logger.info("Console task started.")

        # Start digital input reading task thread accordingly
        if args.enable_digital_input:
            pedal_thread = threading.Thread(target=read_digital_input_task, args=(teleop,), daemon=True)
            pedal_thread.start()
            logger.info("DI reading task started.")

        console_thr.join()

        # Set the stop event when console thread exits
        _stop_event.set()  
        
        if pedal_thread:
            pedal_thread.join(timeout=1.0)

        # Stop teleop process 
        teleop.Stop()
        logger.info("Teleop stopped.")
        
    except KeyboardInterrupt:
        logger.info("Program interrupted by user")
        _stop_event.set()
    except Exception as e:
        logger.error(f"Exception in main: {e}")
        _stop_event.set()
        sys.exit(1)
    finally:
        # Ensure threads are joined properly
        if console_thr and console_thr.is_alive():
            console_thr.join(timeout=1.0)
        if pedal_thread and pedal_thread.is_alive():
            pedal_thread.join(timeout=1.0)
        
        # Ensure teleop is stopped
        if teleop:
            try:
                teleop.Stop()
            except:
                pass  # Ignore errors during cleanup


if __name__ == "__main__":
    main()



